### PR TITLE
Rules

### PR DESCRIPTION
**Pull request by BrianBerzins**
Opening this PR despite knowing that there are still a couple things that must be cleaned up (such as CHANGELOG and README) so I can start getting some feedback.

This is an extensive refactor/rewrite that aims to correct two things missing in the original proof of concept:
1. modular design/extensibility
2. testability

Major changes: 
### Rules
- Rules have been abstracted and moved to their own package. 
- The `Rule` interface defines two methods.
  - `load() (bool, error)` for determining if a rule should be loaded and used
  - `shouldReap() (bool, string)` for determining if a rule is flagging a pod for deletion (with explaination)
- Recommended practice is to load rules based on presence/values of environment variables

### Testability
- Sever waves of refactoring we done to allow for significantly greater test isolation
- Overall behavior should still be tested in a live system (either via minikube to an otherwise safe to fail testing environment before being pushed out).
- Full integration tests have not yet been automated, but I'm playing around with some ideas for this.

### CODE REVIEW

File: rules/chaos.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	chance, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		return false, fmt.Errorf("invalid chaos chance %s", err)
+	}
+	rule.chance = chance
+	return true, nil
+}
+
+func (rule *chaos) ShouldReap(pod v1.Pod) (bool, string) {
+	return rand.Float64() < rule.chance, "was flagged for chaos"
```
Comment (GregoryDosh) [inline rules/chaos.go]: We should probably set the seed to something given that by default the Seed will revert to a value of 1.
https://golang.org/pkg/math/rand/#Seed

Something like `time.Now().UnixNano()` might by simple enough for now.

File: minikube-deploy.sh [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#!/usr/bin/env bash
```
Comment (slushpupie) [inline minikube-deploy.sh]: Any reason to force bash here?  Script seems simple enough to just use `#!/bin/sh`

File: rules/chaos.go (hunk 1 above)
Comment (slushpupie) [inline rules/chaos.go]: Yeah, we are not doing anything crypto/secure related, we just want a random distribution for events. Current time for seeding should be safe.

File: minikube-deploy.sh (hunk 2 above)
Comment (BrianBerzins) [inline minikube-deploy.sh]: checking quickly, but I think you're correct. IDE added it as bash. (incoming update)

**Comment by BrianBerzins** [comment]
Resorting to begging for feedback on overall design/usability/etc...

**Comment by slushpupie** [comment]
Overall I like the split on rules.  Some example configs might help understand how to use it.

File: main.go [hunk 3]
```diff
@@ -26,81 +22,53 @@ func clientSet() *kubernetes.Clientset {
...
-	requirement, err := labels.NewRequirement(options.excludeLabelKey, selection.NotEquals, excludeValueSet)
-	if err != nil {
-		panic(err.Error())
+func getPods(clientSet *kubernetes.Clientset, options options) *v1.PodList {
+	coreClient := clientSet.CoreV1()
+	pods := coreClient.Pods(options.namespace)
+	listOptions := v1.ListOptions{}
+	if options.labelExclusion != nil {
+		selector := labels.NewSelector().Add(*options.labelExclusion)
+		listOptions = v1.ListOptions{LabelSelector: selector.String()}
```
Comment (GregoryDosh) [inline main.go]: Why not change the field of the previous struct instead? `listOptions.LabelSelector = selector.String()`

Comment (BrianBerzins) [inline main.go]: because I am the dumb :P will simplify it :)

**Pull request merged**